rfd/0199-hardware-key-agent.md

```diff
@@ -0,0 +1,588 @@
+with the key through the hardware key agent that is already running.
+
+#### Dependent clients
+
+Dependent clients should interact with the hardware key agent seamlessly so
+that there is limited UX degradation when compared to the a client connecting
+directly to the hardware key.
+
+##### Agent stops running
+
+If the agent stops unexpectedly during a dependent client's operation, it may
+lead to an error for the dependent client. Rather than returning this error,
+the client will log the error as debug and try again by connecting directly to
+the hardware key.
+
+#### Hardware Key prompts
+
+The agent is [responsible for hardware key PIN and touch prompts](#pin-and-touch-prompts).
+As a result, the dependent client will hang while it is prompted and handled
+in the agent client. Teleport Connect will foreground these prompts, so it
+should be clear to the user how to complete the prompts and proceed with the
+dependent client.
+
+Future improvement: instead of leaving the prompt purely up to the agent client,
+the agent client could propagate the prompt to the dependent client requesting
+a signature through a bi-directional, streaming version of the `Sign` rpc. The
+dependent could then prompt for PIN or touch like normal, e.g. in the terminal.
+
+### Proto
+
+#### `HardwareKeyAgentService`
+
+```proto
+// HardwareKeyAgentService provides an agent service for hardware key (PIV) signatures.
+// This allows multiple Teleport clients to share a PIV connection rather than blocking
+// each other, due to the exclusive nature of PIV connections. This also enables shared
+// hardware key states, such as a custom PIN cache shared across Teleport clients.
+service HardwareKeyAgentService {
+  // Sign the given digest with the specified hardware private key. If a hash or salt
+  // was used to produce the digest, HashName and SaltLength must be provided as well.
+  //
+  // This rpc implements Go's crypto.Signer interface.
+  rpc Sign(SignRequest) returns (Signature) {}
+  // GetInfo gets info for the specified hardware private key, including the public key,
+  // attestation statement, and supported pin and touch policies.
+  rpc GetInfo(GetInfoRequest) returns (GetInfoResponse) {}
+}
+
+// SignRequest is a request to perform a signature with a specific hardware private key.
+message SignRequest {
+  // KeyRef references a specific hardware private key.
+  KeyRef key_ref = 1;
+  // PublicKeyDer is a public key encoded in PKIX, ASN.1 DER form.
+  // This is provided to verify that the hardware private key has not
+  // been swapped out against the public key expected by the client.
+  bytes public_key_der = 2;
+  // Digest is a hashed message to sign.
+  bytes digest = 3;
+  // HashName is the name of the hash used to generate the digest.
+  HashName hash_name = 4;
+  // SaltLength determines the length of the salt added to the digest before a signature.
+  // Only required for PSS signatures.
+  oneof salt_length {
```

Comment:
I'm going to double check whether we need PSS signatures and thus the salt length. In the past we did for (some?) TLS handshakes, but that may have changed since https://github.com/gravitational/teleport/blob/master/rfd/0136-modern-signature-algorithms.md.

If we do need PSS signatures, I'll change this to just be a precomputed salt length - https://github.com/go-piv/piv-go/blob/v1.11.0/third_party/rsa/pss.go#L146